## **Eduardo Carrasco Vidal** <img src="img/logo.png" align="right" style="width: 120px;"/>

**Doctorado en Ingeniería de Sistemas Complejos, Universidad Adolfo Ibáñez.**

**Profesor:** Gonzalo Ruz Heredia.
__Ayudante:__ Carlos Sepúlveda Morales.

**Curso:** Machine Learning (Aprendizaje Automático).

Enlace al repositorio del alumno en [GitHub](https://github.com/educarrascov/DISC_MachineLearning) _@educarrascov_

![Python](https://img.shields.io/badge/python-%2314354C.svg) 

## Instrucciones
* La tarea es individual
* Por favor responder en este mismo notebook
* Renombrar el archivo agregando su nombre, por ejemplo tarea3_Contreras_Almada.ipynb
* Subir el archivo al link de entrega Trabajo en webcursos que será habilitado
* __Fecha de entrega:__ viernes 7 de octubre, 23:55 hrs.

# Tarea Nº 3: Machine Learning.

Considere el dataset CallCenterData de la [publicación](https://www.sciencedirect.com/science/article/abs/pii/S0957417412004605) vista en clase.

In [10]:
# se efectúa la carga de la base de datos
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib as plt 
from matplotlib.pyplot import figure
df = pd.read_csv('data/CallCenterData.csv', sep=',')

In [11]:
df.shape

(660, 11)

In [12]:
df.info() #verificar cantidad de columnas/filas, tipos de variables.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660 entries, 0 to 659
Data columns (total 11 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   ID                          660 non-null    int64  
 1   Edad                        660 non-null    int64  
 2   Sexo                        660 non-null    object 
 3   NivelEducacion              660 non-null    object 
 4   EstadoCivil                 660 non-null    object 
 5   NivelSocEco                 660 non-null    object 
 6   HorasLogeadas               660 non-null    float64
 7   HorasHabladas               660 non-null    float64
 8   ContactosEfectivosPromedio  660 non-null    float64
 9   RegistrosTerminados         660 non-null    float64
 10  ProdMayorUbral              660 non-null    object 
dtypes: float64(4), int64(2), object(5)
memory usage: 56.8+ KB


In [13]:
conti_data = df.iloc[:,[1,6,7,8,9]].copy()
categ_data = df.iloc[:,[2,3,4,5]].copy()

In [14]:
conti_data

Edad  HorasLogeadas  HorasHabladas  ContactosEfectivosPromedio  \
0      32       151.1525        45.8525                  192.250000   
1      24        71.1800        24.7600                    1.000000   
2      29       120.5600        49.4500                  156.000000   
3      48        90.8500        28.8500                  442.000000   
4      25       190.1800        41.1600                    0.000000   
..    ...            ...            ...                         ...   
655    50       193.7820        90.1750                  376.500000   
656    28       193.3200        71.5200                  390.000000   
657    29       120.8200        39.1300                   85.666667   
658    28       164.6262        51.7097                  428.600000   
659    33       121.2960        35.7900                  147.200000   

     RegistrosTerminados  
0             591.750000  
1             501.000000  
2             993.000000  
3             569.000000  
4               0.000000  
..                   ...  
655           908.000000  
656           687.500000  
657           471.666667  
658           494.600000  
659           404.000000  

[660 rows x 5 columns]

In [19]:
conti_data.iloc[:,0]

0      32
1      24
2      29
3      48
4      25
       ..
655    50
656    28
657    29
658    28
659    33
Name: Edad, Length: 660, dtype: int64

In [22]:
X_disc = disc.transform(conti_data)

NotFittedError: This KBinsDiscretizer instance is not fitted yet. Call 'fit' with appropriate arguments before using this estimator.

In [9]:
a = pd.X_disc

AttributeError: module 'pandas' has no attribute 'X_disc'

Como se observa, el detalle de las variables es el siguiente:
1. **Numéricas:** `ID`, `Edad`, `HorasLogeadas`, `HorasHabladas`, `ContactosEfectivosPromedio`, `RegistrosTerminados`.
2. **Categóricas:** `Sexo`, `NivelEducacion`, `EstadoCivil`, `NivelSocEco`, `ProdMayorUbral`.

Todas tienen la misma cantidad de datos (660), sin presentar ninguno nulo.

In [ ]:
df.head()

In [ ]:
df.columns

In [ ]:
df.describe(include='all')

Convierta los atributos categóricos a númericos

In [ ]:
# find categorical variables
categorical = [var for var in df.columns if df[var].dtype=='O']
print('There are {} categorical variables\n'.format(len(categorical)))
print('The categorical variables are :\n\n', categorical)

In [ ]:
df[categorical].head()

In [ ]:
# check missing values in categorical variables
df[categorical].isnull().sum()

In [ ]:
# check for cardinality in categorical variables
for var in categorical:  
    print(var, ' contiene ', len(df[var].unique()), ' etiquetas')

In [ ]:
# find numerical variables
numerical = [var for var in df.columns if df[var].dtype!='O']
print('There are {} numerical variables\n'.format(len(numerical)))
print('The numerical variables are :', numerical)

In [ ]:
df[numerical].head()

In [ ]:
# check missing values in numerical variables
df[numerical].isnull().sum()

In [ ]:
X = df.drop(['ProdMayorUbral','ID'], axis=1)

In [ ]:
X.shape

In [ ]:
X.info()

In [ ]:
def get_ohe(X):
    temp_df = pd.DataFrame(data=ohe.transform(X[['Sexo','NivelEducacion','EstadoCivil','NivelSocEco']]), columns=ohe.get_feature_names_out())
    X.drop(columns=['Sexo','NivelEducacion','EstadoCivil','NivelSocEco'], axis=1, inplace=True)
    X = pd.concat([X.reset_index(drop=True), temp_df], axis=1)
    return X

In [ ]:
from sklearn.preprocessing import OneHotEncoder
ohe = OneHotEncoder(handle_unknown='ignore', sparse=False, dtype='int')
ohe.fit(X[['Sexo','NivelEducacion','EstadoCivil','NivelSocEco']])
X = get_ohe(X)

In [ ]:
X.shape

In [ ]:
X.head(5)

In [ ]:
X.info()

Codifique la variable de salida ProdMayorUbral, utilice: etiqueta __si__ codificada como 1, y etiqueta __no__ como 0.

In [ ]:
from sklearn.preprocessing import LabelEncoder
lb = LabelEncoder()
df['ProdMayorUbral'] = lb.fit_transform(df['ProdMayorUbral'])
y = df['ProdMayorUbral']

In [ ]:
y.head(5)

Genere una partición 70% para entrenamiento y el 30% restante para prueba

In [ ]:
# split X and y into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 218)

In [ ]:
# check the shape of X_train and X_test
X_train.shape, X_test.shape

In [ ]:
X_train

In [ ]:
from sklearn.preprocessing import MinMaxScaler

scaling = MinMaxScaler()
X_train = scaling.fit_transform(X_train)
X_test = scaling.transform(X_test)

In [ ]:
X_train = pd.DataFrame(X_train)

In [ ]:
X_test = pd.DataFrame(X_test)

In [ ]:
X_train.head()

In [ ]:
col_names = X.columns
X_train.columns = X.columns

In [ ]:
X_train.head()

## Pregunta 1
Entrene un clasificador naive Bayes usando sólo los atributos socio-demográficos. Luego evalúe en el conjunto de prueba generando el classification report.

### Respuesta:

In [ ]:
X_train1 = X_train.iloc[:,[0,5,6,7,8,9,10,11,12,13,14,15,16,17]].copy()
X_test1 = X_test.iloc[:,[0,5,6,7,8,9,10,11,12,13,14,15,16,17]].copy()

In [ ]:
X_train1

In [ ]:
X_train1.info()

In [ ]:
from sklearn.naive_bayes import BernoulliNB
model_sociodem = BernoulliNB()
score = []
model_sociodem.fit(X_train1, y_train)

In [ ]:
from sklearn.metrics import accuracy_score,r2_score as r2
from sklearn.metrics import classification_report, plot_confusion_matrix

predicciones_sociodem = model_sociodem.predict(X_test1)
accuracy_sociodem = accuracy_score(
            y_true    = y_test,
            y_pred    = predicciones_sociodem,
            normalize = True
           )
print(f"El porcentaje de clasificaciones correctas para el NB de Variables Socio Demográficas es: {round(100 * accuracy_sociodem,2)} %")
score.append(round(100 * accuracy_sociodem,2))

In [ ]:
predicciones_sociodem

In [ ]:
print('Training set score: {:.4f}'.format(model_sociodem.score(X_train1, y_train)))
print('Test set score: {:.4f}'.format(model_sociodem.score(X_test1, y_test)))

In [ ]:
print(classification_report(y_test, predicciones_sociodem))

In [ ]:
from sklearn.metrics import confusion_matrix
cf_matrix_1 = confusion_matrix(
            y_true    = y_test,
            y_pred    = predicciones_sociodem)
print(cf_matrix_1)

In [ ]:
import seaborn as sns
group_names = ['True Neg','False Pos','False Neg','True Pos']
group_counts = ["{0:0.0f}".format(value) for value in
                cf_matrix_1.flatten()]
group_percentages = ["{0:.2%}".format(value) for value in
                     cf_matrix_1.flatten()/np.sum(cf_matrix_1)]
labels = [f"{v1}\n{v2}\n{v3}" for v1, v2, v3 in
          zip(group_names,group_counts,group_percentages)]
labels = np.asarray(labels).reshape(2,2)
sns.heatmap(cf_matrix_1, annot=labels, fmt='', cmap='Greys')

## Pregunta 2
Lo mismo que la pregunta anterior pero usando sólo los atributos operacionales. Mejora el desempeño? argumente en función de las métricas de desempeño.

### Respuesta:

In [ ]:
X_train2 = X_train.iloc[:,[1,2,3,4]].copy()
X_test2 = X_test.iloc[:,[1,2,3,4]].copy()

In [ ]:
X_train2

In [ ]:
from sklearn.naive_bayes import GaussianNB
model_operacionales = GaussianNB()

score = []
model_operacionales.fit(X_train2,y_train)

In [ ]:
predicciones_operacionales = model_operacionales.predict(X_test2)
accuracy_operacionales = accuracy_score(
            y_true    = y_test,
            y_pred    = predicciones_operacionales,
            normalize = True
           )
print(f"El porcentaje de clasificaciones correctas para el NB de Variables Operacionales es: {round(100 * accuracy_operacionales,2)} %")
score.append(round(100 * accuracy_operacionales,2))

In [ ]:
print(classification_report(y_test, predicciones_operacionales))

In [ ]:
from sklearn.metrics import confusion_matrix
cf_matrix_2 = confusion_matrix(
            y_true    = y_test,
            y_pred    = predicciones_operacionales)
print(cf_matrix_2)

In [ ]:
import seaborn as sns
group_names = ['True Neg','False Pos','False Neg','True Pos']
group_counts = ["{0:0.0f}".format(value) for value in
                cf_matrix_2.flatten()]
group_percentages = ["{0:.2%}".format(value) for value in
                     cf_matrix_2.flatten()/np.sum(cf_matrix_2)]
labels = [f"{v1}\n{v2}\n{v3}" for v1, v2, v3 in
          zip(group_names,group_counts,group_percentages)]
labels = np.asarray(labels).reshape(2,2)
sns.heatmap(cf_matrix_2, annot=labels, fmt='', cmap='Greys')

## Pregunta 3
Lo mismo que antes pero ahora considerando todos los atributos. Mejora el desempeño? Valdrá la pena usar todos los atributos realmente? argumente en función de las métricas de desempeño.



### Respuesta

In [ ]:
from sklearn.naive_bayes import GaussianNB
model_NB = GaussianNB()

score = []
model_NB.fit(X_train,y_train)

In [ ]:
predicciones_NB = model_NB.predict(X_test)
accuracy_NB = accuracy_score(
            y_true    = y_test,
            y_pred    = predicciones_NB,
            normalize = True
           )
print(f"El porcentaje de clasificaciones correctas para el NB de todas las variables es: {round(100 * accuracy_NB,2)} %")
score.append(round(100 * accuracy_NB,2))

In [ ]:
print(classification_report(y_test, predicciones_NB))

In [ ]:
from sklearn.metrics import confusion_matrix
cf_matrix_3 = confusion_matrix(
            y_true    = y_test,
            y_pred    = predicciones_NB)
print(cf_matrix_3)

In [ ]:
import seaborn as sns
group_names = ['True Neg','False Pos','False Neg','True Pos']
group_counts = ["{0:0.0f}".format(value) for value in
                cf_matrix_3.flatten()]
group_percentages = ["{0:.2%}".format(value) for value in
                     cf_matrix_3.flatten()/np.sum(cf_matrix_3)]
labels = [f"{v1}\n{v2}\n{v3}" for v1, v2, v3 in
          zip(group_names,group_counts,group_percentages)]
labels = np.asarray(labels).reshape(2,2)
sns.heatmap(cf_matrix_3, annot=labels, fmt='', cmap='Greys')

## Pregunta 4
Qué ocurre si entrena un árbol de decisión usando todos los atributos. Luego evalúe en el conjunto de prueba generando el classification report. Mejora el desempeño? argumente en función de las métricas de desempeño.
Qué atributos aparecen en el árbol resultante? 

### Respuesta

## Pregunta 5
Entrene un clasificador naive Bayes usando los atributos que aparecen en el árbol de decisión.  Luego evalúe en el conjunto de prueba generando el classification report. Mejora el desempeño al comparar los resultados de las preguntas 1, 2, y 3? argumente en función de las métricas de desempeño.

### Respuesta

## Suerte!